From 3c5e0e3ee57bb8c0214598e55f336cd1fd4f92ca Mon Sep 17 00:00:00 2001
From: Noam Camus <ncamus@habana.ai>
Date: Tue, 12 Nov 2019 03:27:59 +0200
Subject: [PATCH] [SW-6379] goya2: crt0 and linker script per binary

Binaries arcboot_rom and arcboot starts their execution differently
The first starts after CPU waken out of reset hence needs IVT.
The second is called by first by simple jump call hence start entry
point should be known e.g. right at binary begin.

Change-Id: I8b253ecd0682eadb8ffc8d7cad8f55c9c3d05ff0
---

diff --git a/Makefile b/Makefile
index b166293..19eebf9 100644
--- a/Makefile
+++ b/Makefile
@@ -18,7 +18,7 @@
 
 LDFLAGS = -flto
 LDFLAGS += -L $(LIBDIR) -lsb -lellipsys
-LDFLAGS += -T arcboot.lds
+LDFLAGS += -T $@.lds
 LDFLAGS += -nostartfiles
 
 # This combination allows removing unused symbols from libsb.a and libellipsys.a
@@ -38,7 +38,7 @@
 
 arcboot: arcboot.o crt0.o
 	$(CC) $^ $(LDFLAGS) -o $@
-arcboot_rom: arcboot_rom.o crt0.o
+arcboot_rom: arcboot_rom.o crt0_rom.o
 	@$(CC) $^ $(LDFLAGS) -o $@
 
 %.o: %.c
diff --git a/arcboot.lds b/arcboot.lds
index 17abac3..ce4ef86 100644
--- a/arcboot.lds
+++ b/arcboot.lds
@@ -20,7 +20,6 @@
 	DCCM	: ORIGIN = 0x80000000, LENGTH = 16K
 }
 
-REGION_ALIAS("REGION_IVT",	ICCM)
 REGION_ALIAS("REGION_TEXT",	ICCM)
 REGION_ALIAS("REGION_DATA",	FWSRAM)
 REGION_ALIAS("REGION_SDATA",	DCCM)
@@ -33,10 +32,6 @@
 
 SECTIONS
 {
-	.ivt : {
-		KEEP (*(.ivt));
-	} > REGION_IVT
-
 	.text : {
 		crt0.o (.text*)
 		*(.text)
diff --git a/arcboot_rom.lds b/arcboot_rom.lds
new file mode 100644
index 0000000..847b11f
--- /dev/null
+++ b/arcboot_rom.lds
@@ -0,0 +1,93 @@
+/*
+ * (C) Copyright 2017 HabanaLabs, Ltd.
+ * Noam Camus <noamikong@gmail.com>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+OUTPUT_FORMAT("elf32-littlearc", "elf32-bigarc", "elf32-littlearc")
+OUTPUT_ARCH(arc)
+
+MEMORY
+{
+	ICCM	: ORIGIN = 0x00000000, LENGTH = 64K
+	SRAM	: ORIGIN = 0x10000000, LENGTH = 128M
+	STM	: ORIGIN = 0x18000000, LENGTH = 32M
+	SPI	: ORIGIN = 0x1A000000, LENGTH = 16M
+	SBROM	: ORIGIN = 0x1B040000, LENGTH = 64K
+	FWSRAM	: ORIGIN = 0x1B050000, LENGTH = 64K
+	CS	: ORIGIN = 0x1C000000, LENGTH = 64M
+	DCCM	: ORIGIN = 0x80000000, LENGTH = 16K
+}
+
+REGION_ALIAS("REGION_IVT",	ICCM)
+REGION_ALIAS("REGION_TEXT",	ICCM)
+REGION_ALIAS("REGION_DATA",	FWSRAM)
+REGION_ALIAS("REGION_SDATA",	DCCM)
+REGION_ALIAS("REGION_SBSS",	DCCM)
+REGION_ALIAS("REGION_BSS",	DCCM)
+REGION_ALIAS("REGION_HEAP",	DCCM)
+REGION_ALIAS("REGION_STACK",	DCCM)
+
+ENTRY(__start)
+
+SECTIONS
+{
+	.ivt : {
+		KEEP (*(.ivt));
+	} > REGION_IVT
+
+	.text : {
+		crt0_rom.o (.text*)
+		*(.text)
+	} > REGION_TEXT
+
+	.data : {
+		*(.data)
+	} > REGION_DATA
+
+	.sdata : {
+		__SDATA_BEGIN__ = .;
+		*(.sdata .sdata.* .gnu.linkonce.s.*)
+		_edata = .;
+		PROVIDE (edata = .);
+	} > REGION_SDATA
+
+	.sbss : {
+		__bss_start = .;
+		PROVIDE (__sbss_start = .);
+		PROVIDE (___sbss_start = .);
+		*(.dynsbss)
+		*(.sbss .sbss.* .gnu.linkonce.sb.*)
+		*(.scommon)
+		PROVIDE (__sbss_end = .);
+		PROVIDE (___sbss_end = .);
+	} > REGION_SBSS
+
+	.bss : {
+		*(.dynbss)
+		*(.bss .bss.* .gnu.linkonce.b.*)
+		*(COMMON)
+		/* Align here to ensure that the .bss section occupies space up to
+		_end. Align after .bss to ensure correct alignment even if the
+		.bss section disappears because there are no input sections. */
+		. = ALIGN(32 / 8);
+		_end = .;
+		PROVIDE (end = .);
+	} > REGION_BSS
+
+	.heap : {
+		. = ALIGN(32 / 8);
+		__start_heap = . ;
+		. = . + (DEFINED(__HEAP_SIZE) ? __HEAP_SIZE : (DEFINED(__DEFAULT_HEAP_SIZE) ? __DEFAULT_HEAP_SIZE : 1k)) ;
+		__end_heap = . ;
+		. = ALIGN(0x8);
+	} > REGION_HEAP
+
+	.stack : {
+		. = ALIGN(32 / 8);
+		__stack = . ;
+		. = . + (DEFINED(__STACK_SIZE) ? __STACK_SIZE : (DEFINED(__DEFAULT_STACK_SIZE) ? __DEFAULT_STACK_SIZE : 512)) ;
+		__stack_top = . ;
+	} > REGION_STACK
+}
\ No newline at end of file
diff --git a/crt0.S b/crt0.S
index 87a3b76..ae22f47 100644
--- a/crt0.S
+++ b/crt0.S
@@ -14,43 +14,6 @@
 
 .extern main
 
-.section .ivt, "a", @progbits
-
-; Helper macro to define weak symbols to include into interrupt vector table.
-; User code may define those functions in them, so user function will be
-; referenced in the IVT. By default all handlers point to _exit_halt - so they
-; always cause application halt, because if application causes an exception or
-; interrupt, but doesn't set a handler for it - something is wrong in
-; application. Exception is "start" entry of IVT, which points to __start
-; function.
-#define IVT_ENTRY(name) \
-    .word name `\
-    .weak name `\
-    .set name, _exit_halt
-
-; handler's name,		number,	name,			offset in IVT (hex/dec)
-.word __start			; 0	program entry point	0x0		0
-IVT_ENTRY(memory_error)		; 1	memory_error		0x4		4
-IVT_ENTRY(instruction_error)	; 2	instruction_error	0x8		8
-IVT_ENTRY(EV_MachineCheck)	; 3	EV_MachineCheck		0xC		12
-IVT_ENTRY(EV_TLBMissI)		; 4	EV_TLBMissI		0x10		16
-IVT_ENTRY(EV_TLBMissD)		; 5	EV_TLBMissD		0x14		20
-IVT_ENTRY(EV_ProtV)		; 6	EV_ProtV		0x18		24
-IVT_ENTRY(EV_PrivilegeV)	; 7	EV_PrivilegeV		0x1C		28
-IVT_ENTRY(EV_SWI)		; 8	EV_SWI			0x20		32
-IVT_ENTRY(EV_Trap)		; 9	EV_Trap			0x24		36
-IVT_ENTRY(EV_Extension)		; 10	EV_Extension		0x28		40
-IVT_ENTRY(EV_DivZero)		; 11	EV_DivZero		0x2C		44
-IVT_ENTRY(EV_DCError)		; 12	EV_DCError		0x30		48
-IVT_ENTRY(EV_Maligned)		; 13	EV_Maligned		0x34		52
-IVT_ENTRY(EV_Ex14)		; 14	unused			0x38		56
-IVT_ENTRY(EV_Ex15)		; 15	unused			0x3C		60
-IVT_ENTRY(IRQ_Timer0)		; 16	Timer 0			0x40		64
-IVT_ENTRY(IRQ_Timer1)		; 17	Timer 1			0x44		68
-IVT_ENTRY(IRQ_18)		; 18				0x48		72
-IVT_ENTRY(IRQ_19)		; 19				0x4C		76
-IVT_ENTRY(IRQ_20)		; 20				0x50		80
-
 .section .text.__startup, "ax", @progbits
 
 ;; lf2flt: to avoid compiler (buildroot wrapper) warning during linkage
diff --git a/crt0_rom.S b/crt0_rom.S
new file mode 100644
index 0000000..87a3b76
--- /dev/null
+++ b/crt0_rom.S
@@ -0,0 +1,103 @@
+;; *******************************************************************************************************************
+;; -----------------
+;; ARC start up file
+;; -----------------
+;; The startup code for the ARC family of processors does the following
+;; before transferring control to user defined main label :
+;;	1. Set sp to __stack_top (link time variable)
+;;	2. Set fp to zero
+;;	3. Zero out the bss section (for uninitialized globals)
+;; After returning from main, the processor is halted and the pipeline is flushed out
+;;
+;; *******************************************************************************************************************
+.file "crt0.S"
+
+.extern main
+
+.section .ivt, "a", @progbits
+
+; Helper macro to define weak symbols to include into interrupt vector table.
+; User code may define those functions in them, so user function will be
+; referenced in the IVT. By default all handlers point to _exit_halt - so they
+; always cause application halt, because if application causes an exception or
+; interrupt, but doesn't set a handler for it - something is wrong in
+; application. Exception is "start" entry of IVT, which points to __start
+; function.
+#define IVT_ENTRY(name) \
+    .word name `\
+    .weak name `\
+    .set name, _exit_halt
+
+; handler's name,		number,	name,			offset in IVT (hex/dec)
+.word __start			; 0	program entry point	0x0		0
+IVT_ENTRY(memory_error)		; 1	memory_error		0x4		4
+IVT_ENTRY(instruction_error)	; 2	instruction_error	0x8		8
+IVT_ENTRY(EV_MachineCheck)	; 3	EV_MachineCheck		0xC		12
+IVT_ENTRY(EV_TLBMissI)		; 4	EV_TLBMissI		0x10		16
+IVT_ENTRY(EV_TLBMissD)		; 5	EV_TLBMissD		0x14		20
+IVT_ENTRY(EV_ProtV)		; 6	EV_ProtV		0x18		24
+IVT_ENTRY(EV_PrivilegeV)	; 7	EV_PrivilegeV		0x1C		28
+IVT_ENTRY(EV_SWI)		; 8	EV_SWI			0x20		32
+IVT_ENTRY(EV_Trap)		; 9	EV_Trap			0x24		36
+IVT_ENTRY(EV_Extension)		; 10	EV_Extension		0x28		40
+IVT_ENTRY(EV_DivZero)		; 11	EV_DivZero		0x2C		44
+IVT_ENTRY(EV_DCError)		; 12	EV_DCError		0x30		48
+IVT_ENTRY(EV_Maligned)		; 13	EV_Maligned		0x34		52
+IVT_ENTRY(EV_Ex14)		; 14	unused			0x38		56
+IVT_ENTRY(EV_Ex15)		; 15	unused			0x3C		60
+IVT_ENTRY(IRQ_Timer0)		; 16	Timer 0			0x40		64
+IVT_ENTRY(IRQ_Timer1)		; 17	Timer 1			0x44		68
+IVT_ENTRY(IRQ_18)		; 18				0x48		72
+IVT_ENTRY(IRQ_19)		; 19				0x4C		76
+IVT_ENTRY(IRQ_20)		; 20				0x50		80
+
+.section .text.__startup, "ax", @progbits
+
+;; lf2flt: to avoid compiler (buildroot wrapper) warning during linkage
+.global	lf2flt
+lf2flt:
+
+.global __start
+.type __start, @function
+.align 4
+
+; Startup code for the ARC600, ARC700 and ARCv2 processors
+; NOTE:  The following restrictions apply on zero overhead loops (other
+; restrictions are not pertinent to this code)
+; - loop end should be 4 instruction words away from the lp_count setting
+;   instruction
+; - loop body should have at least two instruction words
+__start:
+	mov gp, @__SDATA_BEGIN__
+
+	mov_s r2, @__sbss_start	; r2 = start of the bss section
+	sub r3, @_end, r2	; r3 = size of the bss section in bytes
+	; set up the loop counter register to the size (in words) of the bss section
+	asr_s r13, r3
+	asr.f lp_count, r13
+
+	mov sp, @__stack_top	; initialize stack pointer
+	mov_s r3, 0
+	; loop to zero out the bss.  Enter loop only if lp_count != 0
+	lpnz @.Lend_zbss
+	st.ab r3, [r2, 4]	; zero out the word
+	nop
+.Lend_zbss:
+
+	;; branch to main
+	bl.d @main
+	mov fp, 0		; initialize frame pointer
+				; r0 contains exit code
+	j @_exit_halt
+
+.global _exit_halt
+.type _exit_halt, @function
+.align 4
+_exit_halt:
+				; r0 contains exit code
+	flag 0x01
+	nop
+	nop
+	nop
+	j @_exit_halt
+	nop
